glommio/src/io/dma_file_stream.rs

```diff
@@ -1325,6 +1325,10 @@ impl<'a> AsyncWrite for &'a DmaStreamWriter {
         buf: &[u8],
     ) -> Poll<io::Result<usize>> {
         let mut state = self.state.borrow_mut();
+        if state.waker.is_some() {
```

Comment:
This doesn't seem right in two ways. First, I believe you're supposed to always grab the waker from the provided Context if you're pending. If you just grabbed the first one, it's a recipe for never seeing the wake event. Similarly, once this gets into the polling state once, I think it remains stuck that way because the poll will always see a waker.

Reply:
Also, I would recommend adding a test to repro the failure in the issue so that we know it's fixed.

Reply:
I think the correct fix is closer to changing the code in `add_waker` to not assert that the waker `is_none` since that's not guaranteed (you may enter the poll method multiple times AFAIK).

Reply:
Maybe we need to store a `Vec` of `Waker`s? although I'm afraid it will blow up when polled a lot...

I'm not familiar enough with async rust yet, I think I'll take the time to learn it a bit more, and then think of a better solution.

Reply:
Ok I think I got to the conclusion that it is not too interesting to check for parallel writes, but I would like to hear your opinion of course.

Anyway, I updated the PR to simply remove the assert for now.

Reply:
I think the first thing is to come up with a minimal repro as a test that obviously should work and doesn’t and explain the specific sequence of events that leads to a problem (ie is this a bug in the code due to receiving a spurious wake or something else).

Reply:
Also, you mentioned keeping a a vec of Wakers. FWIW, that's not something you need to do unless you're actually managing a collection of futures. The Waker you get on a poll supersedes & invalidates all previous wakers (i.e. once you get polled, you MUST stop using the previous waker unless it's the same wake event). This confused me as well as the documentation here isn't quite so clear.

There's a `will_wake` method you can use to check if it's the same underlying Waker which can be a fast-path to avoid cloning the Waker. That's not particularly important for Glommio since the Waker's it's dealing with are single-threaded reference counts which are essentially free to clone / unref so replacing the waker blindly is fine (i.e. 2 extra integer additions are nothing in the scheme of handling the poll event). The `will_wake` optimization is more valuable if you're polling a huge number of Wakers in a runtime like tokio where it's using atomic reference counts.

As for how to track the parallel writer thing, I might explore something simple like setting a dirty bit in the write entrypoint and then clearing it when Poll::Ready is returned (& if the write entrypoint encounters the dirty bit, return an error). All that being said, I'm not the maintainer of this library, not at all familiar with DmaFileStream, & haven't really investigated how the code works / what it's intent is.

Reply:
I think it's a bit dangerous to just remove the assert, without a reproducer. If you get polled twice, you can consume the buffer twice and end up with a corrupted file.

A panic is better than a corrupted file, so we should at least understand why this is happening with a bit more certainty.

Reply:
You were, right, I tried this change in the program I'm working on, and some writes were simply gone (which makes sense looking at the code).

I've uploaded a better fix, which after trying out again with my program, no writes are gone and the issue https://github.com/DataDog/glommio/issues/610 program is no longer crashing.

Reply:
Ok I thought of a better solution, where it will also fix https://github.com/DataDog/glommio/pull/612 in one go!